### PR TITLE
runner: mount raw block device

### PR DESCRIPTION
**Pull request by skyzh**
This pull request adds support for mounting a raw block device to the VM for future local file cache implementation.

* Raw block device does not have a filesystem UUID, and therefore we cannot use `blkid` to find the disk. We attached the disk with `serial=<ID>`.
* Due to (probably) a bug of QEMU, we cannot directly append `serial=<ID>` in the `-drive` field. It can only be specified in `-device`. And it doesn't boot if we only change one drive to use `-device virtio`. We have to change all attached drives to use `-drive` + `-device`.
* `blkid` does not look at serial number. `udevadm info --query=all` will only print the serial number after `udevadm trigger`, but it doesn't work at boot. Therefore, we have to use `udevadm -a` to walk through all attributes and find the serial number.
* If the serial number in block device information matches the disk ID, we will create a soft link to `mountPath`.

https://github.com/neondatabase/neon/issues/6919

### CODE REVIEW

File: neonvm/runner/main.go [hunk 1]
```diff
@@ -277,11 +281,19 @@ func createISO9660runtime(
 				mounts = append(mounts, fmt.Sprintf(`/neonvm/bin/mount %s $(/neonvm/bin/blkid -L %s) %s`, opts, disk.Name, disk.MountPath))
 				// Note: chmod must be after mount, otherwise it gets overwritten by mount.
 				mounts = append(mounts, fmt.Sprintf(`/neonvm/bin/chmod 0777 %s`, disk.MountPath))
+				diskCnt += 1
+			case disk.RawDisk != nil:
+				device := fmt.Sprintf("/dev/vd%s", string(rune(int('a')+diskCnt)))
```
Comment (sharnoff) [inline neonvm/runner/main.go:286]: This seems fragile (smells like multiple sources of truth). Any ideas for ways around it?

File: neonvm/runner/main.go [hunk 2]
```diff
@@ -703,16 +728,31 @@ func main() {
...
+			extraOptions = "," + extraOptions
+		}
+		if deviceExtraOptions != "" && deviceExtraOptions[0] != ',' {
+			deviceExtraOptions = "," + deviceExtraOptions
+		}
+		qemuCmd = append(qemuCmd, "-drive", fmt.Sprintf("id=%s,file=%s,if=none,index=%d%s", id, file, *diskIdxCnt, extraOptions))
+		qemuCmd = append(qemuCmd, "-device", fmt.Sprintf("virtio-blk,drive=%s%s", id, deviceExtraOptions))
+		*diskIdxCnt += 1
+		return qemuCmd
+	}
```
Comment (sharnoff) [inline neonvm/runner/main.go:744]: Would be nice to have this as a standalone function if possible (perhaps: return only the args that would be added, and then `qemuCmd = append(qemuCmd, qemuDiskArgs(..)...)` ?)

Comment (sharnoff) [reply neonvm/runner/main.go:744]: It'd also be great to have this as a separate commit (and then merge via rebase onto main), or as a separate PR and have this one based on it. That way, it's clear which changes are refactoring(ish) vs adding the feature.

**Review by sharnoff — COMMENTED** [review]
partial review; focusing on the larger stuff first.

File: neonvm/runner/main.go (hunk 1 above)
Comment (skyzh) [reply neonvm/runner/main.go:286]: Yep it's true, but I could not find a good way of doing an 1v1 mapping between serial number and `vd*`. The way it works is that we first guess where the raw block device would be, i.e., vde, and then use the utility tools to verify if it matches the serial number.

File: neonvm/runner/main.go (hunk 2 above)
Comment (skyzh) [reply neonvm/runner/main.go:744]: sounds good I'll make it in a separate commit.

**Comment by skyzh** [comment]
TODO: add test cases, and ready for review :)

File: neonvm/runner/main.go (hunk 1 above)
Comment (Omrigan) [reply neonvm/runner/main.go:286]: How about with smth like `lsblk -o "serial,name" -p`? Or perhaps look it up how `lsblk` does it and reimplement the same way.

File: neonvm/runner/main.go [hunk 3]
```diff
@@ -558,6 +583,22 @@ func runInitScript(logger *zap.Logger, script string) error {
 	return nil
 }
 
+func qemuDiskArgs(id string, file string, extraOptions string, deviceExtraOptions string, diskIdxCnt *int) []string {
+	if extraOptions != "" && extraOptions[0] != ',' {
+		extraOptions = "," + extraOptions
```
Comment (Omrigan) [inline neonvm/runner/main.go:588]: if pass it as a map - will it be prettier? 

Or use builder here.

**Comment by Omrigan** [comment]
> Due to (probably) a bug of QEMU, we cannot directly append serial=<ID> in the -drive field. It can only be specified in -device. And it doesn't boot if we only change one drive to use -device virtio. We have to change all attached drives to use -drive + -device.

Perhaps this is intentional. From `man qemu-system-x86_64`:

```
       -drive option[,option[,option[,...]]]
              Define  a new drive. This includes creating a block driver node (the backend) as well
              as a guest device, and is mostly a shortcut for defining the corresponding  -blockdev
              and -device options.
```

So the idea is `-blockdev` is responsible for the disk "backend" and `-device` is for disk "frontend". Since `serial` is a virtualized property, it is define on a "frontend" side.

File: neonvm/runner/main.go [hunk 4]
```diff
@@ -558,6 +583,22 @@ func runInitScript(logger *zap.Logger, script string) error {
...
+	}
+	if deviceExtraOptions != "" && deviceExtraOptions[0] != ',' {
+		deviceExtraOptions = "," + deviceExtraOptions
+	}
+
+	args := []string{}
+	args = append(args, "-drive", fmt.Sprintf("id=%s,file=%s,if=none,index=%d%s", id, file, *diskIdxCnt, extraOptions))
```
Comment (Omrigan) [inline neonvm/runner/main.go:595]: maybe, we shoud use `-blockdev` here, if we have `-device` also? 

See also https://github.com/neondatabase/autoscaling/pull/822#issuecomment-1986842294

Comment (skyzh) [reply neonvm/runner/main.go:595]: i'll have a try!

**Comment by skyzh** [comment]
> How about with smth like lsblk -o "serial,name" -p? Or perhaps look it up how lsblk does it and reimplement the same way.

@Omrigan unfortunately, lsblk does not show the serial number or name for the raw block device :( could be a bug with lsblk.

File: neonvm/runner/main.go [hunk 5]
```diff
@@ -277,11 +281,19 @@ func createISO9660runtime(
 				mounts = append(mounts, fmt.Sprintf(`/neonvm/bin/mount %s $(/neonvm/bin/blkid -L %s) %s`, opts, disk.Name, disk.MountPath))
 				// Note: chmod must be after mount, otherwise it gets overwritten by mount.
 				mounts = append(mounts, fmt.Sprintf(`/neonvm/bin/chmod 0777 %s`, disk.MountPath))
+				diskCnt += 1
+			case disk.RawDisk != nil:
+				device := fmt.Sprintf("/dev/vd%s", string(rune(int('a')+diskCnt)))
+				// If the serial matches the record, soft-link it to the mountPath.
+				mounts = append(mounts, fmt.Sprintf(`(/neonvm/bin/udevadm info --query=all --name=%s -a | grep 'ATTR{serial}==%q') && /neonvm/bin/ln -s %s %s && /neonvm/bin/chmod 0777 %s`, device, disk.Name, device, disk.MountPath, disk.MountPath))
```
Comment (shayanh) [inline neonvm/runner/main.go:288]: nit: I suggest moving the `fmt.Sprintf` to a separate line to make it easier to read.

File: neonvm/runner/main.go [hunk 6]
```diff
@@ -558,6 +583,22 @@ func runInitScript(logger *zap.Logger, script string) error {
...
+	if extraOptions != "" && extraOptions[0] != ',' {
+		extraOptions = "," + extraOptions
+	}
+	if deviceExtraOptions != "" && deviceExtraOptions[0] != ',' {
+		deviceExtraOptions = "," + deviceExtraOptions
+	}
+
+	args := []string{}
+	args = append(args, "-drive", fmt.Sprintf("id=%s,file=%s,if=none,index=%d%s", id, file, *diskIdxCnt, extraOptions))
+	args = append(args, "-device", fmt.Sprintf("virtio-blk,drive=%s%s", id, deviceExtraOptions))
```
Comment (shayanh) [inline neonvm/runner/main.go:596]: Why are we adding a device here? I thought it might be necessary only for block devices but here we add it for all the drives. Previously we only had `-drive` for disks.

File: neonvm/runner/main.go (hunk 4 above)
Comment (shayanh) [inline neonvm/runner/main.go:595]: I wonder what is the effect of `index` since we didn't use it everywhere before.

File: neonvm/runner/main.go [hunk 7]
```diff
@@ -558,6 +583,22 @@ func runInitScript(logger *zap.Logger, script string) error {
 	return nil
 }
 
+func qemuDiskArgs(id string, file string, extraOptions string, deviceExtraOptions string, diskIdxCnt *int) []string {
```
Comment (shayanh) [inline neonvm/runner/main.go:586]: Nit: having an int pointer and mutating it doesn't look good. Also, increasing the `diskIdxCnt` seems out of scope for this function. To avoid this you can define a closure in main where it calls `qemuDiskArgs` and then increases the counter, which is locally defined there.

File: neonvm/runner/main.go (hunk 6 above)
Comment (skyzh) [reply neonvm/runner/main.go:596]: if we don't do it for all devices, it does not boot unfortunately 🤣

File: neonvm/runner/main.go (hunk 4 above)
Comment (skyzh) [reply neonvm/runner/main.go:595]: added it for an extra layer of stability of assigning `vdX`, should be fine if we don't have that.

File: neonvm/runner/main.go (hunk 7 above)
Comment (skyzh) [reply neonvm/runner/main.go:586]: sounds good :)

**Comment by skyzh** [comment]
Put this pull request on hold until we hit the disk space limit. This pull request only makes sense if the LFC take a significant amount of space and we need to reclaim the space very quickly. Also need to address the page cache problem -- it seems that raw block devices do not take the full page cache.

**Comment by Omrigan** [comment]
Discussed with @skyzh, closing for the time being.
